Microsoft.Research/ManagedContract.Setup/buildMSI10.xml

```diff
@@ -370,8 +371,33 @@
 
     ==========================================================-->
   <Target Name="MergeExes"
-    DependsOnTargets="ILMergeFoxtrot;ILMergeAsmMeta;ILMergeClousot;ILMergeCCDoc"/>
+    DependsOnTargets="PrepareMerges;ExecuteMerges"/>
 
+  <Target Name="PrepareMerges" BeforeTargets="ExecuteMerges"
+    DependsOnTargets="RestoreToolPackages;PrepareILMerge"/>
+	
+  <Target Name="RestoreToolPackages">
+    <Exec Command="&quot;$(SolutionDir).NuGet\NuGet.exe&quot; restore &quot;$(SolutionDir)packages.config&quot; -PackagesDirectory &quot;$(SolutionDir)packages&quot;" />

```

Comment:
OK, I think I understand our different point of view:

Current CC build script is similar to the one presented in my [link](https://docs.nuget.org/consume/package-restore/team-build#the-general-approach). In both cases MsBuild is executed twice:
- First one on `build.proj` in example and `buildMSI10.xml` in CC.
- The second one on `BingSearcher.sln` in example and `CodeContracts.sln` in CC.

This first msbuild run acts like a bat file from your point of view. In both cases nuget is being called before second msbuild run. This is a solution proposed in the NuGet link I provided. And it is stated that this solution differs from a legacy one (the one which uses Nuget.targets) in which:
"The previous implementation had a chicken-and-egg problem for packages that want to extend the build process because NuGet restored packages while building the project"
The presented solution does not have this kind of problems because MsBuild for which the packages were downloaded has not been run yet.

So behavior described in your point 1. is exactly the same for my approach.
All the arguments from point 2 can also be applied to all the logic contained in buildMSI10.xml.
I admit, current project file is quite complex and hard to maintain. We should rethink how the build process should be defined. Then I can agree to move *all the logic* from `buildMSI10.xml` to sh1 script. Bat file may be too clumsy.

Moving only `nuget.exe restore` call to a bat file complicates things. Current approach centralizes all the build activities keeping them in one file. When you move nuget.exe call to a bat file moving another activity from an xml to bat before nuget may not be so simple (depends on complexity of the activity to be moved).